Give objects the ability to fine-tune petsc options

pbehne opened the pull request:
Currently, petsc options are handled in the `MOOSE::PetscSupport` namespace in `PetscSupport.C`. The functions defined here set petsc options for various general cases. However, these general settings are not applicable to some cases and can result in unused petsc option warmings.

This PR is the first step to address #28053. A capability is introduced such that various moose objects can specify exceptions to the general cases in `PetscSupport`. These exceptions are given as a list of options that are to be ignored, or they would otherwise cause unused option warnings. With this design, special cases that cause warnings can be addressed by ignoring the irrelevant petsc options. An example of how to use this new features is given in `ExplicitTimeIntegrator`. With this change, the unused option warnings that appeared when running `test/tests/time_integrators/actually_explicit_euler_verification/ee-1d-linear.i` disappear.

If we want to go with this proposed design, the test suite should be examined for all occurrences of unused option warnings and these options should be turned off for those tests. This examination can be easily conducted by setting a default value for the `absent_out` parameter in the `RunApp` tester to identify relevant tests that have unused option warnings.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2287080/) on 5f29419 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27948/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2287095/) on 5f29419 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>67db2d</th>
      <th colspan=4>#27948 5f2941</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.06%</td>
      <td>85.06%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104409</td>
      <td>104408</td>
      <td>-1</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18336</td>
      <td>18337</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27948/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27948/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/xfem)

</details>



This comment will be updated on new commits.

pbehne committed:
Changed sub_pc_factor_shift_type to pc_factor_shift_type since the former was an unused option in this test.

pbehne committed:
Silence petsc unused options if printUsage is called.

pbehne committed:
Added a version of addValidName that takes another MultiMooseEnum as a parameter.

pbehne committed:
Method renames

processPetscPairs -> addPetscPairsToPetscOptions
processPetscFlags -> addPetscFlagsToPetscOptions

pbehne committed:
Split storePetscOptions into 4 functions

The following new funcitons were created to lend clarity to what the code does:
setSolverTypeFromParams
setLineSearchFromParams
setMFFDTypeFromParams
storePetscOptionsFromParams

pbehne committed:
Split getCommonPetscFlags/Keys into smaller funcs

that use new methods to get common ksp and snes-related flags/keys.

pbehne committed:
Added necessary structure to PetscOptions class to track options to be ignored.

pbehne committed:
Making ExplicitTimeIntegrators ingnore snes options.

pbehne committed:
Finished initial interface for ignored options, cleanup

Renamed PetscOptions’ ‘ignored_flags’ attribute to ‘ignored_options’.
Created a wrapped version of setSinglePetscOption called
setSinglePetscOptionIfNotIgnored that will only set the option if it is not in
the ignored list.
Modified setSolverOptions to use this wrapper so ignored options aren’t blindly
set.
Removed the disableIgnoredPetscFlags function because it didn’t play well with
user-specified pets options from the input file.
Renamed disableCommonKSP(SNES)Flags to ignoreCommonKSP(SNES)Options and modified
the logic to add these options to the ignored list instead of removing them from
the flags and pairs attributes because nothing prevents these options from being
re-added to flags and pairs later.
Cleaned up ExplicitTimeIntegrator to use ignoreCommonSNESOptions.
Create new attribute ‘user_set_options` in PetscOptions to keep track of which
options are set by the user at the input file. These options do not get ignored,
even if they are in the ignored_options list.

Refs #28053

pbehne committed:
Method name formatting

The 'AddPetscFlagsToPetscOptions' and 'AddPetscPairsToPetscOptions' were wrongly
capitalized. This has been corrected. Furthermore, code in the moose modules
that called these functions under their old names has been updated.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2605827/), step Generate coverage on 29c1110 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>38207e</th>
      <th colspan=4>#27948 29c111</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.25%</td>
      <td>85.25%</td>
      <td>-0.00%</td>
      <td>89.36%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>107993</td>
      <td>108046</td>
      <td>+53</td>
      <td>84</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18682</td>
      <td>18692</td>
      <td>+10</td>
      <td>10</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27948/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27948/coverage/framework)


## Modules coverage

Mutiple base shas were found when retrieving head reports:
  4e4e5dc3b1066150df33a7fcd81a7aecfcfff6a9
  38207e829fa6e8719263c8ab300526f41a2bc77e

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/framework)
- `chemical_reactions` (FAILED)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/combined)
- `contact` (FAILED)
- `electromagnetics` (FAILED)
- `external_petsc_solver` (FAILED)
- `fluid_properties` (FAILED)
- `fsi` (FAILED)
- `functional_expansion_tools` (FAILED)
- `geochemistry` (FAILED)
- `heat_transfer` (FAILED)
- `level_set` (FAILED)
- `misc` (FAILED)
- `navier_stokes` (FAILED)
- `optimization` (FAILED)
- `peridynamics` (FAILED)
- `phase_field` (FAILED)
- `porous_flow` (FAILED)
- `ray_tracing` (FAILED)
- `rdg` (FAILED)
- `reactor` (FAILED)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27948/coverage/richards)
- `scalar_transport` (FAILED)
- `solid_mechanics` (FAILED)
- `solid_properties` (FAILED)
- `stochastic_tools` (FAILED)
- `thermal_hydraulics` (FAILED)
- `xfem` (FAILED)

</details>

## Warnings

- `framework` new line coverage rate 89.36% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2605827/), step Verify coverage on 29c1110 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 84.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 81.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 88.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2605827/), step Verify coverage on 29c1110 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 84.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 81.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 88.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

pbehne commented on framework/src/parser/CommandLine.C:
```diff
@@ -414,6 +416,10 @@ CommandLine::printUsage() const
+
+  // If we get here, we are not running a simulation and should silence petsc's unused options
+  // warning
+  Moose::PetscSupport::setSinglePetscOption("-options_left", "0");
```
Changes in this file are unrelated to the new `ignored_options` attribute in `PetscOptions`, but having this here does get rid of some additional unused option warnings.

pbehne commented on framework/src/timeintegrators/ExplicitTimeIntegrator.C:
```diff
@@ -53,6 +53,9 @@ ExplicitTimeIntegrator::ExplicitTimeIntegrator(const InputParameters & parameter
 
   if (_solve_type == LUMPED || _solve_type == LUMP_PRECONDITIONED)
     _ones = addVector("ones", false, PARALLEL);
+
+  // ignore all common SNES-related petsc options to prevent unused option warnings
+  Moose::PetscSupport::ignoreCommonSNESOptions(_fe_problem);
```
This change shows how the new `ignored_options` attribute in `PetscOptions` is utilized to disable options specific to the `ExplicitTimeIntegrator` that are not used.

pbehne commented:
@lindsayad, would you mind reviewing this PR to ensure that this new feature is how we want to handle unused petsc options? Confirmation on this would be good to have before I hunt down all occurrences of the warnings in the test suite. If not you, could you tag someone you think would be a good reviewer?

I'm thinking we split issue #28053 up. After merging this PR, we can have another PR that suppresses unused options in the test suite.

lindsayad commented on framework/include/utils/PetscSupport.h:
```diff
@@ -38,14 +38,23 @@ namespace PetscSupport
   /// Single value PETSc options (flags)
   MultiMooseEnum flags;
 
+  /// Flags to explicitly not set, even if they are specified in flags
+  MultiMooseEnum ignored_options;
```
I would prefer that we generally replace `ignored` with something like `dont_add_these`. Ignoring something scares me. It sounds like someone has asked for something but we're going to ignore it. "Don't add this" sounds a lot less scary and also reflects more truly what your code is doing: it's making it so that we don't add particular flags that don't make sense for a given simulation as opposed to "we're going to add these flags that don't make sense and then ignore them"

lindsayad commented on framework/include/utils/PetscSupport.h:
```diff
@@ -153,6 +196,15 @@ void setSinglePetscOption(const std::string & name,
+Same as setSinglePetscOption, but does not set the option if it belongs to the given list of
+ignored options.
+*/
+void setSinglePetscOptionIfNotIgnored(const MultiMooseEnum & ignored_options,
```
```suggestion
void setSinglePetscOptionIfAppropriate(const MultiMooseEnum & ignored_options,
```
or something like that

lindsayad commented on framework/include/utils/PetscSupport.h:
```diff
@@ -153,6 +196,15 @@ void setSinglePetscOption(const std::string & name,
                           FEProblemBase * const problem = nullptr);
 
+/**
+Same as setSinglePetscOption, but does not set the option if it belongs to the given list of
+ignored options.
```
```suggestion
Same as setSinglePetscOption, but does not set the option if it doesn't make sense for the current simulation type
```

lindsayad commented on framework/include/utils/PetscSupport.h:
```diff
@@ -187,6 +244,16 @@ void disableNonlinearConvergedReason(FEProblemBase & fe_problem);
+/**
+ * disable common ksp flags
+ */
+void ignoreCommonKSPOptions(FEProblemBase & fe_problem);
```
```suggestion
void dontAddCommonKSPOptions(FEProblemBase & fe_problem);
```

lindsayad commented on framework/include/utils/PetscSupport.h:
```diff
@@ -187,6 +244,16 @@ void disableNonlinearConvergedReason(FEProblemBase & fe_problem);
+/**
+ * disable common snes flags
+ */
+void ignoreCommonSNESOptions(FEProblemBase & fe_problem);
```
```suggestion
void dontAddCommonSNESOptions(FEProblemBase & fe_problem);
```

pbehne committed:
Addressed Alex's review.

lindsayad commented on framework/include/utils/PetscSupport.h:
```diff
@@ -177,6 +229,11 @@ void colorAdjacencyMatrix(PetscScalar * adjacency_matrix,
                           const char * coloring_algorithm);
 
+/**
+ * disable a particular petsc flag
```
More particular language would be good. Does this remove a flag if it exists? Does it tell the `PetscOptions` object to not add the flag if someone attempts to add it later? Both?

lindsayad replied:
We use the disable language repeatedly in the following doxygen. I think if you give a clear definition of what "disable" means here, you do not have to go and add it to all the following doc-strings

lindsayad replied:
If "disable" just means "dontAdd", then let's change the function name to match those that you added later in the file

pbehne replied:
Once a petsc option is already actually set, I don't like to mess with it. My approach is to modify the `PetscOptions` object before options are set. So I think it makes sense to change `disable` to `dontAdd`. What do you like more, `dontAdd` or `dontSet`?

lindsayad replied:
Probably `dontAdd`

pbehne commented on framework/include/utils/PetscSupport.h:
```diff
@@ -102,23 +111,45 @@ PetscErrorCode petscLinearMonitor(KSP /*ksp*/, PetscInt its, PetscReal rnorm, vo
+/**
+ * Stores the Petsc flags and pair options fron the input params in the given PetscOptions object.
+ */
+void storePetscOptionsFromParams(PetscOptions & po,
```
Consider the current definition of `storePetscOptions` (which I haven't pushed just yet with the updated signatures):
```
void
storePetscOptions(FEProblemBase & fe_problem, const InputParameters & params)
{
  setSolveTypeFromParams(fe_problem, params);
  setLineSearchFromParams(fe_problem, params);
  setMFFDTypeFromParams(fe_problem, params);
  storePetscOptionsFromParams(fe_problem, params);
}
```
See that I split `storePetscOptions` into 4 subfunctions. I did this because it felt a bit more organized to me. Note that while `storePetscOptionsFromParams` *stores* options in a `PetscOptions` object, the other 3 subfunctions *set* options directly and don't use the `PetscOptions` object. Now that I think about it more, however, these 3 subfunctions are just storing options, there are just dedicated attributes for the storage of solve type, line search, and mffd type outside of a `PetscOptions` object.

Would you prefer I eliminate the subfunctions and just have one long `storePetscOptions` function?

lindsayad replied:
Hmm Im starting to realize that this PR is going to have a lot of conflicts with #29072. One way we could remove conflict is if you adopt https://github.com/idaholab/moose/pull/29072/files#diff-813f6769dc958bcf657d6f3422f2bec7ca9d01f4df25a8fbab2905ffbec8fa7bR116. I believe that function is equivalent to the sum of `setSolveTypeFromParams`, `setLineSearchFromParams`, and `setMFFDTypeFromParams`. Then you can simplify the existing `storePetscOptions` function like I did in #29072. 

It would be nice to avoid adding the `storePetscOptionsFromParams` function because as it stands now a developer may think that it does the same thing as `storePetscOptions` when in reality they differ in setting/not-setting solve-type/line-search/mffd-type

lindsayad replied:
I am fine with keeping `setSolveTypeFromParams`, `setLineSearchFromParams`, and `setMFFDTypeFromParams` and having those be called from `Moose::PetscSupport::processSingletonMooseWrappedOptions`

pbehne replied:
Whoa. There are a lot of conflicts. Let me know if my latest updates are what you wanted, or if you want more than that.

pbehne committed:
Addressed Alex's review 2.0

pbehne committed:
Addressed Alex's review 2.5: incorporating processSingletonMooseWrappedOptions.

pbehne commented:
@lindsayad, pinging you again on this so it stays on your list.

zachmprince commented on framework/include/utils/PetscSupport.h:
```diff
@@ -38,14 +38,23 @@ namespace PetscSupport
+  /// Flags to explicitly not set, even if they are specified programmatically
+  MultiMooseEnum dont_add_these_options;
+
+  /// Options that are set by the user at the input level
+  MultiMooseEnum user_set_options;
```
There doesn't seem to be anything testing the fact that a warning gets published on a user-defined option. So I would suggest adding something where 1) a nonsense flag is set and 2) a flag that is usually added programmatically, isn't, but is added via input.

pbehne replied:
See latest changes. I added a test to cover 2). 1) seems to be already be covered by `misc/petsc_option_left.test_non_petsc_options_not_left`.

zachmprince reviewed (commented):
Implementation looks good. I just had one suggestion for a test or two. I'm not sure if there is something already out there, though.

pbehne committed:
Added test (Addressed Zach's review)

Added a test to ensure that a unused petsc option set by the user is not ignored
and emits a warning.

lindsayad commented on framework/include/utils/PetscSupport.h:
```diff
@@ -38,14 +38,23 @@ namespace PetscSupport
   /// PETSc key-value pairs
   std::vector<std::pair<std::string, std::string>> pairs;
 
   /// Single value PETSc options (flags)
   MultiMooseEnum flags;
 
+  /// Flags to explicitly not set, even if they are specified programmatically
+  MultiMooseEnum dont_add_these_options;
+
+  /// Options that are set by the user at the input level
+  MultiMooseEnum user_set_options;
```
To be honest I hate that these are MultiMooseEnums since the PETSc options database is constantly changing and there is no hope of enumerating them all in MOOSE. Additionally, I think a `std::set` or `std::vector` would just be a lot clearer. But I am ok with you adding these new data members, following the existing pattern. Fixing this can be follow-on work by someone

lindsayad commented on framework/include/utils/PetscSupport.h:
```diff
@@ -153,6 +200,15 @@ void setSinglePetscOption(const std::string & name,
 
+/**
+ * Same as setSinglePetscOption, but does not set the option if it doesn't make sense for the
+ * current simulation type.
```
```suggestion
 * current simulation type, e.g. if \p name is contained within \p dont_add_these_options
```

lindsayad commented on framework/include/utils/PetscSupport.h:
```diff
@@ -178,14 +234,34 @@ void colorAdjacencyMatrix(PetscScalar * adjacency_matrix,
 
 /**
- * disable printing of the nonlinear convergence reason
+ * Function to ensure that particular petsc option is not added to the PetscOptions
```
```suggestion
 * Function to ensure that a particular petsc option is not added to the PetscOptions
```

lindsayad commented on framework/src/utils/PetscSupport.C:
```diff
@@ -467,10 +503,14 @@ storePetscOptions(FEProblemBase & fe_problem, const InputParameters & params)
   if (params.isParamValid("line_search"))
   {
-    MooseEnum line_search = params.get<MooseEnum>("line_search");
+    const MooseEnum line_search = params.get<MooseEnum>("line_search");
```
```suggestion
    const auto & line_search = params.get<MooseEnum>("line_search");
```

lindsayad commented on framework/src/utils/PetscSupport.C:
```diff
@@ -494,31 +534,20 @@ storePetscOptions(FEProblemBase & fe_problem, const InputParameters & params)
+{
   if (params.isParamValid("mffd_type"))
   {
     MooseEnum mffd_type = params.get<MooseEnum>("mffd_type");
```
```suggestion
    const auto & mffd_type = params.get<MooseEnum>("mffd_type");
```

lindsayad commented on framework/src/utils/PetscSupport.C:
```diff
@@ -782,32 +825,64 @@ getPetscValidParams()
-      "-ksp_monitor -ksp_monitor_snes_lg-snes_ksp_ew -ksp_snes_ew -snes_converged_reason "
-      "-snes_ksp -snes_ksp_ew -snes_linesearch_monitor -snes_mf -snes_mf_operator -snes_monitor "
-      "-snes_test_display -snes_view",
+      "-ksp_monitor_snes_lg -snes_ksp_ew -ksp_snes_ew -snes_converged_reason "
+      "-snes_ksp -snes_linesearch_monitor -snes_mf -snes_mf_operator -snes_monitor "
+      "-snes_test_display -snes_view -snes_monitor_cancel",
```
yea attempting to keep track of these options is insanity

pbehne replied:
Is there any petsc API that can provide a list of valid options instead of us maintaining these lists?

lindsayad replied:
from @balay:

> I don't think there is a single database to query for all valid options.
> 
> The options are distributed across petsc code - and when that part of the code is run - it checks against the input-options base. [i.e. - if that part of the code is -not- run,there is no check for that option]. Also prefix usage in library or use code can change the options.
> 
> Hence './binary -h' prints only the options that are used/available in  the code path of the currently running code. [and it can't print all]

pbehne replied:
I'm going to mark this as resolved for now to allow the merge. Maybe we will think of a better management system in the future. All of these options were hardcoded in `PetscSupport` before I ever touched it, I just further split them into the ksp and snes categories.

pbehne committed:
Apply suggestions from Alex code review

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

pbehne committed:
Address Alex review

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

lindsayad commented:
@pbehne you've managed to trigger something way down in hypre in the failed next branch testing. It's likely going to take me a decent while to figure it out. In the meantime it's probably best if you rewrite your tests and avoid using hypre

loganharbour commented:
See https://civet.inl.gov/event/217284/

pbehne commented:
> In the meantime it's probably best if you rewrite your tests and avoid using hypre

On it.
